File: Algorithm/Programers_무인도 여행.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;
int n, m;
int mx[4] = { 0,0,1,-1 };
int my[4] = { 1,-1,0,0 };
vector<vector<char>> charMap;
vector<vector<bool>> visited;

int bfs(int _y, int _x) {
	int sum = charMap[_y][_x] - '0';
	visited[_y][_x] = true;
	queue<pair<int, int>> q;
	q.push({ _y,_x });
	while (!q.empty()) {
		int y = q.front().first;
		int x = q.front().second;
		q.pop();
		for (int i = 0; i < 4; i++) {
			int dy = y + my[i];
			int dx = x + mx[i];
			if (dy >= n || dy < 0 || dx >= m || dx < 0) continue;
			if (visited[dy][dx]) continue;
			visited[dy][dx] = true;
			if (charMap[dy][dx] == 'X') continue;
			sum += charMap[dy][dx] - '0';
			q.push({ dy,dx });
		}
	}
	return sum;
}

vector<int> solution(vector<string> maps) {
	n = maps.size();
	m = maps[0].length();
	charMap = vector<vector<char>>(n, vector<char>(m));
	visited = vector<vector<bool>>(n, vector<bool>(m, false));
	vector<int> answer;

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			charMap[i][j] = maps[i][j];
		}
	}
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			if (charMap[i][j] == 'X') continue;
			if (visited[i][j]) continue;
			answer.push_back(bfs(i, j));
		}
	}
	sort(answer.begin(), answer.end());
	if (answer.size() == 0) answer.push_back(-1);

	return answer;
}
//https://school.programmers.co.kr/learn/courses/30/lessons/154540

File: Algorithm/Programers_무인도_여행.java
import java.util.*;

class Solution {
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    static int n, m;
    static boolean[][] visited;
    static int[][] arr;

    public int[] solution(String[] maps) {
        List<Integer>al = new ArrayList<>();
        n = maps.length;
        m = maps[0].length();

        arr = new int[n][m];
        visited = new boolean[n][m];

        for(int i = 0; i < n; i++){
            String str = maps[i];
            for(int j = 0; j < m; j++){
                if(str.charAt(j) != 'X'){
                    arr[i][j] = str.charAt(j) - '0';
                }
            }
        }

        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(arr[i][j] > 0 && !visited[i][j]){
                    al.add(bfs(i, j));
                }
            }
        }

        if(al.size() == 0){
            return new int[] {-1};
        }
        Collections.sort(al);
        int[] answer = new int[al.size()];

        for(int i = 0; i < al.size(); i++){
            answer[i] = al.get(i);
        }

        return answer;
    }

    public int bfs(int r, int c){
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[] {r, c});
        visited[r][c] = true;
        int sum = arr[r][c];

        while(!q.isEmpty()){
            int[] now = q.poll();

            for(int i = 0; i < 4; i++){
                int nr = now[0] + dx[i];
                int nc = now[1] + dy[i];

                if(nr >= n || nc >= m || nr < 0 || nc < 0 || visited[nr][nc] || arr[nr][nc] == 0) continue;

                sum += arr[nr][nc];
                visited[nr][nc] = true;
                q.offer(new int[] {nr, nc});
            }
        }
        return sum;
    }
}

class main{

    public static void main(String[] args) {
        Solution sol = new Solution();
        String[] s ={"X591X","X1X5X","X231X", "1XXX1"};
        System.out.println("sol = " + sol.solution(s));
    }

}
